File: java/java/src/main/java/org/signal/libsignal/crypto/Aes256Ctr32.java
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.crypto;

import org.signal.client.internal.Native;
import org.whispersystems.libsignal.InvalidKeyException;

public class Aes256Ctr32 {
  private final long handle;

  public Aes256Ctr32(byte[] key, byte[] nonce, int initialCtr) throws InvalidKeyException {
    this.handle = Native.Aes256Ctr32_New(key, nonce, initialCtr);
  }

  @Override
  protected void finalize() {
    Native.Aes256Ctr32_Destroy(this.handle);
  }

  public void process(byte[] data) {
    Native.Aes256Ctr32_Process(this.handle, data, 0, data.length);
  }

  public void process(byte[] data, int offset, int length) {
    Native.Aes256Ctr32_Process(this.handle, data, offset, length);
  }

}

File: swift/Sources/SignalClient/Aes256Ctr32.swift
//
// Copyright 2020 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalFfi
import Foundation

public class Aes256Ctr32: ClonableHandleOwner {
    public init<KeyBytes, NonceBytes>(
        _ key: KeyBytes,
        _ nonce: NonceBytes,
        _ initial_ctr: UInt32) throws
        where KeyBytes: ContiguousBytes,
              NonceBytes: ContiguousBytes {

        let handle: OpaquePointer? = try key.withUnsafeBytes { keyBytes in
            try nonce.withUnsafeBytes { nonceBytes in
                var result: OpaquePointer?
                let error = signal_aes256_ctr32_new(&result,
                                                    keyBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                    keyBytes.count,
                                                    nonceBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                    nonceBytes.count,
                                                    initial_ctr)
                try checkError(error)
                return result
            }
        }
        super.init(owned: handle!)
    }

    internal override class func destroyNativeHandle(_ handle: OpaquePointer) -> SignalFfiErrorRef? {
        return signal_aes256_ctr32_destroy(handle)
    }

    public func process<MutableBytes: ContiguousMutableBytes>(_ bytes: inout MutableBytes) throws {
        try bytes.withUnsafeMutableBytes {
            let error = signal_aes256_ctr32_process(nativeHandle,
                                                    $0.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                    $0.count,
                                                    0,
                                                    UInt32($0.count))
            try checkError(error)
        }
    }

    public func process<MutableBytes: ContiguousMutableBytes>(
        _ bytes: inout MutableBytes,
        _ offset: UInt32,
        _ length: UInt32) throws {
        try bytes.withUnsafeMutableBytes {
            let error = signal_aes256_ctr32_process(nativeHandle,
                                                    $0.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                    $0.count,
                                                    offset,
                                                    length)
            try checkError(error)
        }
    }
}
